src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,200 @@
+end
+
+"""
+    PspUpf(Zion::Number, lmax::Number, rgrid::Vector, drgrid::Vector, vloc::Vector,
+           vloc_interp, projs::Vector{Vector{Vector}}, projs_interp, h::Vector{Matrix},
+           pswfcs::Vector{Vector}, pswfc_occs::Vector, rhoatom::Vector, rvlocpzdr::Vector,
+           r2projsdr::Vector{Vector{Vector}}; identifier="")
+
+Construct a Unified Pseudopotential Format pseudopotential. Currently only
+norm-conserving potentials are supported.
+"""
+function PspUpf(Zion, lmax, rgrid::Vector{T}, drgrid, vloc, vloc_interp::IT, projs,
+                projs_interp, h, pswfcs, pswfc_occs, rhoatom, rvlocpzdr, r2projsdr;
+                identifier="") where {T, IT}
+    PspUpf{T,IT}(Zion, lmax, rgrid, drgrid, vloc, vloc_interp, projs, projs_interp, h,
+                 pswfcs, pswfc_occs, rhoatom, rvlocpzdr, r2projsdr, identifier)
+end
+
+function parse_upf_file(path; identifier=path)
+    pseudo = load_upf(path)
+
+    unsupported = []
+    pseudo["header"]["core_correction"]      && push!(unsupported, "non-lin. core correction")
+    pseudo["header"]["has_so"]               && push!(unsupported, "spin-orbit coupling")
+    pseudo["header"]["pseudo_type"] == "SL"  && push!(unsupported, "semilocal potential")
+    pseudo["header"]["pseudo_type"] == "US"  && push!(unsupported, "ultrasoft")
+    pseudo["header"]["pseudo_type"] == "PAW" && push!(unsupported, "plane-augmented wave")
+    pseudo["header"]["has_gipaw"]            && push!(unsupported, "gipaw data")
+    pseudo["header"]["pseudo_type"] == "1/r" && push!(unsupported, "Coulomb")
+    length(unsupported) > 0 && error("Pseudopotential contains the following unsupported" *
+                                     " features/quantities: $unsupported")
+
+    Zion        = Int(pseudo["header"]["z_valence"])     # Pseudo-ion charge
+    rgrid       = pseudo["radial_grid"]                  # Radial grid
+    drgrid      = pseudo["radial_grid_derivative"]       # Integration coeff.s
+    lmax        = pseudo["header"]["l_max"]              # Maximum angular momentum channel
+    vloc        = pseudo["local_potential"] ./ 2         # Local potential (Ry -> Ha)
+    vloc_interp = linear_interpolation((rgrid, ), vloc)  # Interp.or for the local potential
+
+    # Kleinman-Bylander β projectors projs[l+1][i][ir]
+    # NB: UPFs store rβ, not just β
+    projs = map(0:lmax) do l
+        betas_l = filter(beta -> beta["angular_momentum"] == l, pseudo["beta_projectors"])
+        map(beta -> beta["radial_function"] ./ 2, betas_l)  # Ry -> Ha
+    end
+    projs_interp = map(projs) do projs_l
+        map(projs_l) do proj
+            ir_cut = lastindex(proj)
+            linear_interpolation((rgrid[1:ir_cut], ), proj)
+        end
+    end
+
+    # β-projector coupling coefficients h[l+1][i,j] (also called e_{KB}, D_{ij})
+    h = Matrix[]
+    n = 1
+    for l = 0:lmax
+        nproj_l = length(projs[l+1])
+        Dij_l = pseudo["D_ion"][n:n+nproj_l-1, n:n+nproj_l-1] .* 2  # 1/Ry -> 1/Ha
```

Comment:
I do the conversion like this so that the projectors match up with HGH (because the units are Ry * Bohr^{-1/2} in the file). In theory, calculation results are the same either way, and I agree that it's cleaner to treat them as Bohr^{-1/2} and Ry. I can modify the tests and put in an explanation, but it's messy either way.

Reply:
Haven't checked what you settled on, but certainly we want quantities between the HGH and the UPF version to match up.

Reply:
Hm, sucks HGH does the "wrong" thing here... OK to leave it as is then, with a comment